solutions/cf-718C.mdx

```diff
@@ -0,0 +1,176 @@
+
+We will use the term _cycle by k_ to denote transforming some pair $(F_{i - 1}, F_i)$ to $(F_{i - 1 + k}, F_{i + k})$. By default, the term _cycle_ refers to _cycle by 1_.
+
+For each update, we need to cycle each leaf node by some amount $x$ such that the value in a node affected goes from $(F_{v - 1}, F_v)$ to $(F_{v - 1 + x}, F_{v + x})$. We can do this using matrix exponentiation.
+
+First note that for a matrix $\begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}$, we can cycle it by multiplying by $\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}$. That is,
+
+$$
+\begin{pmatrix} F_i & F_{i + 1} \end{pmatrix} = \begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}
+$$
+
+Inductively, we can cycle by $k$ by multiplying multiple times:
+
+$$
+\begin{pmatrix} F_{i - 1 + k} & F_{i + k} \end{pmatrix} = \begin{pmatrix} F_{i - 1} & F_i \end{pmatrix}\begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}^k
+$$
+
+We can use binary exponentiation to quickly compute powers of the two-by-two matrix. 
+
+Secondly, we need to utilize the distribute property of matrix multiplication to same-size matricies. That is, if $m_k$ is a one-by-two matrix for $k \in [1, n]$, then the following property is satisfied:
+
+$$
+(m_1 + m_2 + \dots + m_n) \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} = m_1 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + m_2 \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix} + \dots + m_n \cdot \begin{pmatrix} 0 & 1 \\ 1 & 1 \end{pmatrix}
+$$
+
+In that way, we can utilize lazy propgation on the segment tree. We will store an integer tag in each node denoting the lazy update, which is denotes that we must cycle every node in its subtre by the value of the tag. When propogating, we can simply update the node by binary exponentiating and then multiplying, then storing it in the lazy tag. Querying works as normal on the segment tree. 
+
+## Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define FF first
+#define SS second
+
+using ll = long long;
+using pll = pair<ll, ll>;
+using Matrix = array<array<ll, 2>, 2>;
+
+const int MAXN = 1e5 + 5;
+const ll MOD = 1e9 + 7;
+
+// multiplies two two-by-two matricies
+Matrix multiply(const Matrix &a, const Matrix &b) {
+	return {(a[0][0] * b[0][0] + a[0][1] * b[1][0]) % MOD, 
+		(a[0][0] * b[0][1] + a[0][1] * b[1][1]) % MOD,
+		(a[1][0] * b[0][0] + a[1][1] * b[1][0]) % MOD,
+		(a[1][0] * b[0][1] + a[1][1] * b[1][1]) % MOD};
+}
+
+// multiplies a one-by-two matrix with a two-by-two matrix
+pll multiply(const pll &a, const Matrix &b) {
+	return {(a.FF * b[0][0] + a.SS * b[1][0]) % MOD, 
+		(a.FF * b[0][1] + a.SS * b[1][1]) % MOD};
+}
+
+// returns {0, 1, 1, 1,} to the power of n
+Matrix pow(ll n) { 
+	Matrix res = {1, 0, 0, 1};  // identity matrix
+	Matrix base = {0, 1, 1, 1};  // fibonacci matrix
+	for (; n; n >>= 1) {. // binary exponentiation (log n)
```

Comment:
dustin did you mean to put a period here

Reply:
uuuhhh

Reply:
^^